File: examples/09-stdlib/generated.user.c

#include <math.h>

Float64 clamp(Float64 value, Float64 lo, Float64 hi);
Float64 lerp(Float64 a, Float64 b, Float64 t);
Int32 main(Void);
Str shout(Str value);
Str initials(Str value);

/* module kira:math */
Float64 clamp(Float64 value, Float64 lo, Float64 hi)
{
    return fmax(lo, fmin(value, hi));
}
Float64 lerp(Float64 a, Float64 b, Float64 t)
{
    return (a + ((b - a) * t));
}
/* module app:main */
/* use app:text */
Int32 main(Void)
{
    Str name = "  kira  ";
    Str trimmed = Str_trim(name);
    print("%d\n", Str_length(trimmed));
    print("%s\n", shout(trimmed));
    print("%s\n", initials(trimmed));
    print("%d\n", Str_startsWith(trimmed, "ki"));
    print("%s\n", Str_substring(trimmed, 0, 2));
    Int32 count = 7;
    print("%lld\n", (long long)(((Int64)(count))));
    Set seen = Set_new();
    Set_add(&seen, KIRA_SLOT(1));
    Set_add(&seen, KIRA_SLOT(2));
    Set_add(&seen, KIRA_SLOT(1));
    print("%d\n", Set_size(&seen));
    print("%d\n", Set_contains(&seen, KIRA_SLOT(2)));
    Stack undo = Stack_new();
    Stack_push(&undo, KIRA_SLOT(10));
    Stack_push(&undo, KIRA_SLOT(20));
    Maybe top = Stack_pop(&undo);
    print("%d\n", KIRA_UNSLOT(Int32, Maybe_unwrapOr(&top, KIRA_SLOT(0))));
    Queue jobs = Queue_new();
    Queue_enqueue(&jobs, KIRA_SLOT(1));
    Queue_enqueue(&jobs, KIRA_SLOT(2));
    Maybe next = Queue_dequeue(&jobs);
    print("%d\n", KIRA_UNSLOT(Int32, Maybe_unwrapOr(&next, KIRA_SLOT(0))));
    Map ages = Map_new_s();
    Map_put(&ages, KIRA_SLOT_PTR("ada"), KIRA_SLOT(36));
    Maybe found = Map_get(&ages, KIRA_SLOT_PTR("ada"));
    print("%d\n", Maybe_isSome(&found));
    print("%d\n", KIRA_UNSLOT(Int32, Maybe_unwrapOr(&found, KIRA_SLOT(0))));
    Maybe missing = Map_get(&ages, KIRA_SLOT_PTR("nobody"));
    print("%d\n", KIRA_UNSLOT(Int32, Maybe_unwrapOr(&missing, KIRA_SLOT(-1))));
    List nums = List_new();
    List_add(&nums, KIRA_SLOT(3));
    List_add(&nums, KIRA_SLOT(4));
    print("%d\n", KIRA_UNSLOT(Int32, List_get(&nums, 1)));
    print("%d\n", List_contains(&nums, KIRA_SLOT(3)));
    Maybe absent = Maybe_none();
    print("%d\n", Maybe_isNone(&absent));
    print("%s\n", KIRA_UNSLOT_PTR(Str, Maybe_unwrapOr(&absent, KIRA_SLOT_PTR("fallback"))));
    Maybe present = Maybe_some(KIRA_SLOT_PTR("here"));
    print("%d\n", Maybe_isSome(&present));
    print("%s\n", KIRA_UNSLOT_PTR(Str, Maybe_unwrapOr(&present, KIRA_SLOT_PTR("fallback"))));
    kira_assert((List_size(&nums) == 2), "list should hold two entries");
    print("%s\n", "ok");
    List_dispose(&nums);
    Map_dispose(&ages);
    Queue_dispose(&jobs);
    Stack_dispose(&undo);
    Set_dispose(&seen);
    return 0;
}
/* module app:text */
Str shout(Str value)
{
    return Str_toUpper(value);
}
Str initials(Str value)
{
    return Str_charAt(value, 0);
}

File: examples/09-stdlib/generated.user.js

// module "app:main"
// use "app:text"
function main() {
    const name = "  kira  ";
    const trimmed = kira_str_trim(name);
    kira_trace(kira_str_length(trimmed));
    kira_trace(shout(trimmed));
    kira_trace(initials(trimmed));
    kira_trace(kira_str_startsWith(trimmed, "ki"));
    kira_trace(kira_str_substring(trimmed, 0, 2));
    const count = 7;
    kira_trace(kira_num_toInt64(count));
    const seen = kira_set_new();
    seen.add(1);
    seen.add(2);
    seen.add(1);
    kira_trace(seen.size());
    kira_trace(seen.contains(2));
    const undo = kira_stack_new();
    undo.push(10);
    undo.push(20);
    const top = undo.pop();
    kira_trace(top.unwrapOr(0));
    const jobs = kira_queue_new();
    jobs.enqueue(1);
    jobs.enqueue(2);
    const next = jobs.dequeue();
    kira_trace(next.unwrapOr(0));
    const ages = kira_map_new();
    ages.put("ada", 36);
    const found = ages.get("ada");
    kira_trace(found.isSome());
    kira_trace(found.unwrapOr(0));
    const missing = ages.get("nobody");
    kira_trace(missing.unwrapOr(-1));
    const nums = kira_list_new();
    nums.add(3);
    nums.add(4);
    kira_trace(nums.get(1));
    kira_trace(nums.contains(3));
    const absent = kira_none();
    kira_trace(absent.isNone());
    kira_trace(absent.unwrapOr("fallback"));
    const present = kira_some("here");
    kira_trace(present.isSome());
    kira_trace(present.unwrapOr("fallback"));
    kira_assert((nums.size() == 2), "list should hold two entries");
    kira_trace("ok");
}

// module "app:text"
function shout(value) {
    return kira_str_toUpper(value);
}

function initials(value) {
    return kira_str_charAt(value, 0);
}

main();
